circuits/cpp/src/aztec3/circuits/abis/final_accumulated_data.hpp

```diff
@@ -0,0 +1,201 @@
+using std::is_same;
+
+template <typename NCT> struct FinalAccumulatedData {
+    using fr = typename NCT::fr;
+    using uint32 = typename NCT::uint32;
+    using boolean = typename NCT::boolean;
+    using AggregationObject = typename NCT::AggregationObject;
+
+    AggregationObject aggregation_object{};
+
+    std::array<fr, MAX_NEW_COMMITMENTS_PER_TX> new_commitments{};
+    std::array<fr, MAX_NEW_NULLIFIERS_PER_TX> new_nullifiers{};
+    std::array<fr, MAX_NEW_NULLIFIERS_PER_TX> nullified_commitments{};
+    // For pending nullifiers, we have:
+    // nullifiedCommitments[j] != 0 <==> newNullifiers[j] nullifies nullifiedCommitments[j]
+
+    std::array<fr, MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX> private_call_stack{};
+    std::array<fr, MAX_PUBLIC_CALL_STACK_LENGTH_PER_TX> public_call_stack{};
+    std::array<fr, MAX_NEW_L2_TO_L1_MSGS_PER_TX> new_l2_to_l1_msgs{};
+
+    std::array<fr, NUM_FIELDS_PER_SHA256> encrypted_logs_hash{};
+    std::array<fr, NUM_FIELDS_PER_SHA256> unencrypted_logs_hash{};
+
+    // Here so that the gas cost of this request can be measured by circuits, without actually needing to feed in the
+    // variable-length data.
+    fr encrypted_log_preimages_length = 0;
+    fr unencrypted_log_preimages_length = 0;
+
+    std::array<NewContractData<NCT>, MAX_NEW_CONTRACTS_PER_TX> new_contracts{};
+
+    std::array<OptionallyRevealedData<NCT>, MAX_OPTIONALLY_REVEALED_DATA_LENGTH_PER_TX> optionally_revealed_data{};
+
+    // for serialization, update with new fields
+    MSGPACK_FIELDS(aggregation_object,
+                   new_commitments,
+                   new_nullifiers,
+                   nullified_commitments,
+                   private_call_stack,
+                   public_call_stack,
+                   new_l2_to_l1_msgs,
+                   encrypted_logs_hash,
+                   unencrypted_logs_hash,
+                   encrypted_log_preimages_length,
+                   unencrypted_log_preimages_length,
+                   new_contracts,
+                   optionally_revealed_data);
+    boolean operator==(FinalAccumulatedData<NCT> const& other) const
+    {
+        return aggregation_object == other.aggregation_object && new_commitments == other.new_commitments &&
+               new_nullifiers == other.new_nullifiers && nullified_commitments == other.nullified_commitments &&
+               private_call_stack == other.private_call_stack && public_call_stack == other.public_call_stack &&
+               new_l2_to_l1_msgs == other.new_l2_to_l1_msgs && encrypted_logs_hash == other.encrypted_logs_hash &&
+               unencrypted_logs_hash == other.unencrypted_logs_hash &&
+               encrypted_log_preimages_length == other.encrypted_log_preimages_length &&
+               unencrypted_log_preimages_length == other.unencrypted_log_preimages_length &&
+               new_contracts == other.new_contracts && optionally_revealed_data == other.optionally_revealed_data;
```

Comment:
Would use utils::msgpack_derived_equals` here. With all those struct members it will be significantly less bug prone when doing changes.